File: py-wav/mapper/mapper.py
import json
from json import JSONEncoder
from typing import Dict, Type

from mapper.Mappable import Mappable


def tree_print(msg: str, depth: int):
    print("-" * depth + msg)


def save(source: Mappable, file: str):
    with open(file, 'w') as f:
        json.dump(source, f, cls=SourceEncoder)


class SourceEncoder(JSONEncoder):
    def default(self, o):
        return o.get_params()


class SourceParser:
    def __init__(self, registry: Dict[str, Type[Mappable]]):
        self.registry = registry

    def parse(self, file):
        with open(file) as song:
            raw = json.load(song)
        return self.construct(raw)

    def unpack(self, obj, depth, name="Value"):
        if isinstance(obj, dict):
            return self.construct(obj, depth=depth)
        elif isinstance(obj, list):
            tree_print("unpacking a list", depth)
            return [self.unpack(i, depth=depth+1) for i in obj]
        else:
            tree_print(f"{name}: {obj}", depth)
            return obj

    def construct(self, raw: Dict, depth=0):
        if "type" not in raw:  # this is a data object
            return raw

        type_name = raw.pop("type")
        tree_print(f"Node: {type_name}", depth)

        clazz = self.registry[type_name]

        unpacked = {"type": type_name}

        for key in raw:
            unpacked[key] = self.unpack(raw[key], depth=depth+1, name=key)

        try:
            return clazz(**unpacked)
        except TypeError:
            print(f"Error parsing {type_name}")
            raise

File: web/src/mapper/Mapper.ts
import { MappingBase, MappableData, isMappableData, Mappable } from './Mappable';
import { Registry } from './registry'; 

export class Mapper {
    registry: Registry = null
    quiet = true

    constructor(registry: Registry, quiet: boolean=true) {
        this.registry = registry
        this.quiet = quiet
    }

    tree_print(msg: string, depth: number) {
        if (!this.quiet) {
            console.log("-".repeat(depth) + msg)
        }
    }

    unpack(obj: MappableData | any[] | any, depth: number, name: string="Value")
    : any {
        if (isMappableData(obj)) { 
            return this.construct(obj, depth)
        } else if (Array.isArray(obj)) {
            this.tree_print("array", depth)
            return obj.map(e => this.unpack(e, depth+1))
        } else {
            this.tree_print(name + ": " + obj, depth)
            return obj
        }
    }

    construct(data: MappableData, depth: number=0): Mappable {
        this.tree_print("Node: " + data.type, depth)

        const fields = Object.keys(data)

        const mappedData = Object.fromEntries(fields.map(field => {
            return [field, this.unpack(data[field], depth+1, field)]
        }))

        try {
            const node =  new this.registry[data.type](mappedData)
            return node
        } catch (err) {
            console.log("Error constructing node of type: " + data.type)
            throw err
        }
    }
}